File: src/main/java/com/smartlamp/agent/LlmClient.java
package com.smartlamp.agent;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import tools.jackson.databind.ObjectMapper;
import tools.jackson.databind.node.ArrayNode;
import tools.jackson.databind.node.ObjectNode;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

// 大模型调用封装：兼容 OpenAI 接口协议（OpenAI / DeepSeek / Kimi 等），
// 配置走 application.yml 的 llm: 段（环境变量 LLM_API_KEY / LLM_BASE_URL / LLM_MODEL 占位）
@Component
public class LlmClient {

    private static final int TIMEOUT_MS = 10_000;

    private final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();
    private final ObjectMapper objectMapper = new ObjectMapper();

    @Value("${llm.api-key:}")
    private String apiKey;

    @Value("${llm.base-url:https://api.openai.com/v1}")
    private String baseUrl;

    @Value("${llm.model:}")
    private String model;

    // 未配置 Key 或模型时视为未启用，由调用方降级为本地知识库回答
    public boolean isConfigured() {
        return apiKey != null && !apiKey.isBlank() && model != null && !model.isBlank();
    }

    public String completeChat(String system, String user) {
        if (!isConfigured()) throw new LlmException("LLM not configured");

        String url = baseUrl.replaceAll("/+$", "") + "/chat/completions";

        ObjectNode body = objectMapper.createObjectNode();
        body.put("model", model);
        ArrayNode messages = body.putArray("messages");
        messages.addObject().put("role", "system").put("content", system);
        messages.addObject().put("role", "user").put("content", user);
        body.put("temperature", 0.3);

        try {
            HttpRequest request = HttpRequest.newBuilder(URI.create(url))
                    .timeout(Duration.ofMillis(TIMEOUT_MS))
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer " + apiKey)
                    .POST(HttpRequest.BodyPublishers.ofString(objectMapper.writeValueAsString(body)))
                    .build();
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                throw new LlmException("LLM API 返回 " + response.statusCode());
            }

            String content = objectMapper.readTree(response.body())
                    .path("choices").path(0).path("message").path("content").asText();
            if (content == null || content.isBlank()) {
                throw new LlmException("LLM API 返回内容为空");
            }
            return content.trim();
        } catch (LlmException e) {
            throw e;
        } catch (Exception e) {
            throw new LlmException("LLM API 调用失败: " + e.getMessage());
        }
    }
}

File: backend/agent/llmClient.js
const fs = require("fs");
const path = require("path");

// 与具体厂商解耦：只要接口兼容 OpenAI 协议即可（OpenAI / DeepSeek / Kimi / 通义千问 等）
const ENV_FILE = path.resolve(__dirname, "..", "..", ".env");
const DEFAULT_TIMEOUT_MS = 10000;
let envLoaded = false;

// 简单 .env 加载：只设置尚未定义的环境变量，不覆盖 Shell 中已存在的值
function loadEnvFileOnce() {
  if (envLoaded) return;
  envLoaded = true;

  try {
    if (!fs.existsSync(ENV_FILE)) return;
    const content = fs.readFileSync(ENV_FILE, "utf8");
    for (const line of content.split(/\r?\n/)) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#")) continue;
      const sep = trimmed.indexOf("=");
      if (sep === -1) continue;
      const key = trimmed.slice(0, sep).trim();
      const value = trimmed.slice(sep + 1).trim();
      if (key && process.env[key] === undefined) process.env[key] = value;
    }
  } catch (error) {
    console.warn("[llmClient] 读取 .env 失败:", error.message);
  }
}

function getConfig() {
  loadEnvFileOnce();
  const apiKey = process.env.LLM_API_KEY;
  const model = process.env.LLM_MODEL;
  if (!apiKey || !model) return null;
  return {
    apiKey,
    baseUrl: process.env.LLM_BASE_URL || "https://api.openai.com/v1",
    model,
  };
}

async function completeChat({ system, user }) {
  const config = getConfig();
  if (!config) throw new Error("LLM not configured");

  const url = `${config.baseUrl.replace(/\/+$/, "")}/chat/completions`;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), DEFAULT_TIMEOUT_MS);

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${config.apiKey}`,
      },
      body: JSON.stringify({
        model: config.model,
        messages: [
          { role: "system", content: system },
          { role: "user", content: user },
        ],
        temperature: 0.3,
      }),
      signal: controller.signal,
    });

    if (!response.ok) {
      throw new Error(`LLM API 返回 ${response.status}`);
    }

    const data = await response.json();
    const content = data && data.choices && data.choices[0] && data.choices[0].message && data.choices[0].message.content;
    if (typeof content !== "string" || !content.trim()) {
      throw new Error("LLM API 返回内容为空");
    }
    return content.trim();
  } finally {
    clearTimeout(timer);
  }
}

module.exports = {
  completeChat,
  getConfig,
};
